File: app/tools/network.py
"""Network quick fix actions."""

from __future__ import annotations

from app.core.exec import CommandSpec
from app.tools.base import Action, Tool


class NetworkTool(Tool):
    name = "network"
    description = "Network diagnostics and quick fixes"

    def register_actions(self) -> None:
        self.actions["flush_dns"] = Action(
            id="flush_dns",
            label="Flush DNS",
            description="ipconfig /flushdns",
            exec_steps=[CommandSpec(executor="cmd", command="ipconfig /flushdns", timeout=15)],
        )

        self.actions["winsock_reset"] = Action(
            id="winsock_reset",
            label="Reset Winsock",
            description="netsh winsock reset",
            exec_steps=[
                CommandSpec(
                    executor="cmd",
                    command="netsh winsock reset",
                    timeout=15,
                    elevate=True,
                )
            ],
            requires_admin=True,
        )

        self.actions["renew_ip"] = Action(
            id="renew_ip",
            label="Renew IP",
            description="ipconfig /release && ipconfig /renew",
            exec_steps=[
                CommandSpec(executor="cmd", command="ipconfig /release", timeout=30),
                CommandSpec(executor="cmd", command="ipconfig /renew", timeout=30),
            ],
        )

        self.actions["ping_host"] = Action(
            id="ping_host",
            label="Ping Host",
            description="ping -n 4 <host>",
            exec_steps=[
                CommandSpec(
                    executor="cmd",
                    command="ping -n 4 $env:TARGET_HOST",
                    timeout=20,
                    env={"TARGET_HOST": "8.8.8.8"},
                )
            ],
        )

        self.actions["curl_head"] = Action(
            id="curl_head",
            label="HTTP HEAD",
            description="curl -I <url>",
            exec_steps=[
                CommandSpec(
                    executor="cmd",
                    command="curl -I $env:TARGET_URL",
                    timeout=20,
                    env={"TARGET_URL": "https://www.microsoft.com"},
                )
            ],
        )


__all__ = ["NetworkTool"]

File: rust/src/system/network.rs
//! Network utilities
//!
//! Network diagnostics and quick-fix commands.

use crate::state::{CommandOutput, NetworkAdapter};
use crate::system::command::run_command;
use sysinfo::Networks;


/// Flush DNS cache
pub async fn flush_dns() -> CommandOutput {
    run_command("ipconfig", &["/flushdns"]).await
}

/// Release and renew IP address
pub async fn renew_ip() -> Vec<CommandOutput> {
    let release = run_command("ipconfig", &["/release"]).await;
    let renew = run_command("ipconfig", &["/renew"]).await;
    vec![release, renew]
}

/// Reset Winsock (requires admin)
pub async fn winsock_reset() -> CommandOutput {
    run_command("netsh", &["winsock", "reset"]).await
}

/// Restart HTTP.sys service (requires admin)
/// This clears orphaned HTTP sockets that may be holding ports
pub async fn restart_http_service() -> Vec<CommandOutput> {
    let stop = run_command("net", &["stop", "http", "/y"]).await;
    // Small delay to ensure service fully stops
    tokio::time::sleep(tokio::time::Duration::from_millis(500)).await;
    let start = run_command("net", &["start", "http"]).await;
    vec![stop, start]
}

/// Ping a host
pub async fn ping_host(host: &str) -> CommandOutput {
    run_command("ping", &["-n", "4", host]).await
}

/// HTTP HEAD request
pub async fn http_head(url: &str) -> CommandOutput {
    // Try curl first (modern Windows has it)
    if which::which("curl").is_ok() {
        run_command("curl", &["-I", "--connect-timeout", "10", url]).await
    } else {
        // Fallback to PowerShell
        let script = format!(
            "Invoke-WebRequest -Uri '{}' -Method HEAD -UseBasicParsing | Select-Object -ExpandProperty Headers | ConvertTo-Json",
            url
        );
        run_command("powershell", &["-NoProfile", "-Command", &script]).await
    }
}

/// Get network adapters with their addresses
/// Reserved for future network adapter display feature
#[allow(dead_code)]
pub fn get_network_adapters() -> Vec<NetworkAdapter> {
    let networks = Networks::new_with_refreshed_list();
    
    networks
        .iter()
        .map(|(name, _data)| {
            // sysinfo doesn't give us IP addresses directly in newer versions,
            // so we'll use a simpler approach for now
            NetworkAdapter {
                name: name.to_string(),
                addresses: vec![], // We'll populate this via ipconfig parsing if needed
                is_up: true, // Assume up if listed
            }
        })
        .collect()
}

/// Get detailed network adapter info using ipconfig
/// Reserved for future network adapter display feature
#[allow(dead_code)]
pub async fn get_adapter_details() -> CommandOutput {
    run_command("ipconfig", &["/all"]).await
}
